### PR TITLE
Implement BatchNorm2d

### PR DESCRIPTION
**Pull request by kakao-steve-ai**
## 🙏 Describe the pull request

4D 입력에 대한 BatchNorm2d 구현입니다.

## ✅ Checklist

- [x] Code follows the project's coding conventions and style.
- [x] Tests have been added or updated to cover the changes.
- [x] Documentation has been updated, if necessary.

### CODE REVIEW

File: trident/module.py [hunk 1]
```diff
@@ -126,6 +126,110 @@ def extra_repr(self):
...
+        Returns:
+            an output with the same dimension and shape as an input
+        """
+        assert input.dim() == 4
+        inp = input.view(*input.shape[0:2], -1)
+        assert inp.dim() == 3
+        assert inp.shape[0] > 1
+
+        output = function.batch_norm(
+            inp,
```
Comment (daemyung) [inline trident/module.py]: 인라인 처리 해주세요.

File: tests/test_batch_norm.py [hunk 2]
```diff
@@ -70,9 +70,18 @@ def train(func):
...
     assert output is not None and output.dtype == dtype
+
+
+@pytest.mark.parametrize("num_batches, z_size, y_size, x_size", [(3, 3, 128, 128)])
+def test_batch_norm_2d(num_batches, z_size, y_size, x_size, device, dtype):
+    factory_kwargs = {"device": device, "dtype": dtype}
+    input = torch.randn(num_batches, z_size, y_size, x_size, **factory_kwargs)
+
+    output = trident.BatchNorm2d(z_size, **factory_kwargs).forward(input)
+    assert output is not None and output.dtype == dtype
```
Comment (daemyung) [inline tests/test_batch_norm.py]: `assert`를 `and`로 묶지 말고 2줄로 나눠주세요. 두줄로 작성하면 어떤 조건에 실패하는지 명확해집니다.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: trident/module.py [hunk 3]
```diff
@@ -126,6 +126,110 @@ def extra_repr(self):
...
+        Applies Batch Normalization to an input.
+
+        Args:
+            input: an input
+
+        Returns:
+            an output with the same dimension and shape as an input
+        """
+        assert input.dim() == 4
+        inp = input.view(*input.shape[0:2], -1)
```
Comment (mejai1206) [inline trident/module.py]: `inp`를 쓰신 이유는 `input`을 덮어쓰지 않기 위함인가요?

Comment (kakao-steve-ai) [reply trident/module.py]: 네 `input.shape`를 나중에 쓰기 위해서, 안 덮어쓰려고 이렇게 했습니다.

File: tests/test_batch_norm.py (hunk 2 above)
Comment (kakao-steve-ai) [reply tests/test_batch_norm.py]: 수정하였습니다

File: trident/module.py (hunk 1 above)
Comment (kakao-steve-ai) [reply trident/module.py]: 수정하였습니다

File: trident/module.py [hunk 4]
```diff
@@ -126,6 +126,107 @@ def extra_repr(self):
...
+
+        Args:
+            input: an input
+
+        Returns:
+            an output with the same dimension and shape as an input
+        """
+        assert input.dim() == 4
+
+        output = function.batch_norm(
```
Comment (daemyung) [inline trident/module.py]: 이것도 인라인 처리하면 더 보기 좋겠어요.

File: trident/module.py [hunk 5]
```diff
@@ -126,6 +126,107 @@ def extra_repr(self):
...
+        """
+        Applies Batch Normalization to an input.
+
+        Args:
+            input: an input
+
+        Returns:
+            an output with the same dimension and shape as an input
+        """
+        assert input.dim() == 4
```
Comment (daemyung) [inline trident/module.py]: `assert`말고 예외로 처리 부탁드려요.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: tests/util.py [hunk 6]
```diff
@@ -15,7 +15,7 @@
 import torch
 
 
-def equal(input: torch.Tensor, other: torch.Tensor, eps: float = 1e-02):
+def equal(input: torch.Tensor, other: torch.Tensor, eps: float = 1e-03):
```
Comment (daemyung) [inline tests/util.py]: 잘못 수정된 것 같습니다.

File: trident/module.py (hunk 4 above)
Comment (kakao-steve-ai) [reply trident/module.py]: 수정하였습니다

File: trident/module.py (hunk 5 above)
Comment (kakao-steve-ai) [reply trident/module.py]: 수정하였습니다

File: trident/module.py [hunk 7]
```diff
@@ -80,12 +80,12 @@ def forward(self, input: torch.Tensor):
         """
         if input.dim() == 2:
             inp = input.view(*input.shape, 1)
-        else:
+        elif input.dim() == 3:
             inp = input
-        assert inp.dim() == 3
-        assert inp.shape[0] > 1
+        else:
+            raise ValueError("Input dimension should be 2 or 3.")
```
Comment (daemyung) [inline trident/module.py:82]: 텐서의 모양을 변경하는 코드는 인라인으로 처리 부탁드려요.

**Review by daemyung — CHANGES_REQUESTED** [review]

File: tests/util.py (hunk 6 above)
Comment (kakao-steve-ai) [reply tests/util.py]: 앗 실수로 들어갔네요

File: trident/module.py (hunk 7 above)
Comment (kakao-steve-ai) [reply trident/module.py:82]: 수정하였습니다

**Review by daemyung — APPROVED** [review]
LGTM

**Review by mejai1206 — APPROVED** [review]
LGTM
